tests/backend/checkmach/test_assume_error.ml

```diff
@@ -0,0 +1,225 @@
+(* A1: zero_alloc check fails on [foo] because there is an allocation on the
+   path to exceptional return from the call to [f x].
+   This exception is handled in foo and then [foo] returns normally,
+   making this allocation on a normal return of [foo].
+*)
+module A1 = struct
+  let[@inline never][@local never] h s = [Random.int s ; Random.int s]
+
+  let[@inline never][@local never][@zero_alloc] g x =
+    if x < 0 then Sys.opaque_identity (x+1) else raise (Failure ((string_of_int) x))
+
+  let[@inline never][@local never] handle_exn exn =
+    (match exn with
+     | Failure s -> print_string s; print_newline ()
+     | _ -> failwith "Boo");
+    0
+  ;;
+
+  let[@zero_alloc] f x =
+    match g x with
+    | exception exn ->
+      (handle_exn[@zero_alloc assume never_returns_normally]) exn
+      |> h |> List.hd  (* ignore this allocation *)
+    | answer -> (answer * 2)
+
+  let[@zero_alloc] foo x =
+    try f x with
+    | _ -> 0
+end
+
+(* A2 is a more elaborate version of A1 with more allocations in [foo]
+   that are all correctly reported. *)
+module A2 = struct
+  let[@inline never][@local never] h s = [Random.int s ; Random.int s]
+
+  let[@inline never][@local never][@zero_alloc] g x =
+    if x < 0 then Sys.opaque_identity (x+1) else raise (Failure ((string_of_int) x))
+
+  let[@inline never][@local never] handle_exn exn =
+    (match exn with
+     | Failure s -> print_string s; print_newline ()
+     | _ -> failwith "Boo");
+    0
+  ;;
+
+  let[@zero_alloc] f x =
+    match g x with
+    | exception exn ->
+      (handle_exn[@zero_alloc assume never_returns_normally]) exn
+      |> h |> List.hd  (* ignore this allocation *)
+    | answer -> (answer * 2)
+
+  let[@zero_alloc] foo x =
+    try let y = f x in (x,y) with
+    | _ ->
+      Printf.printf "BOO %x" x;  (* ignore this indirect call *)
+      (x,0)
+end
+
+(* A3: shows that "assume error" causes the check to ignore allocations
+   in [foo] that come after the exceptional return from [f x]
+   and correctly report allocations that happen after normal return
+   from [f x]. *)
+module A3 = struct
+  let[@inline never][@local never] h s = [Random.int s ; Random.int s]
+
+  let[@inline never][@local never][@zero_alloc] g x =
+    if x < 0 then Sys.opaque_identity (x+1) else raise (Failure ((string_of_int) x))
+
+  let[@inline never][@local never] handle_exn exn =
+    (match exn with
+     | Failure s -> print_string s; print_newline ()
+     | _ -> failwith "Boo");
+    0
+  ;;
+
+  let[@zero_alloc] f x =
+    match g x with
+    | exception exn ->
+      (handle_exn[@zero_alloc assume error]) exn
+      |> h |> List.hd  (* ignore this allocation *)
+    | answer -> (answer * 2)
+
+  let[@zero_alloc] foo x =
+    try
+      let y = f x in
+      (x,y)  (* report this allocation *)
+    with
+    | _ ->
+      Printf.printf "BOO %x" x;  (* ignore this indirect call *)
```

Comment:
Why do we want to ignore this call?  It doesn't look like it's on a path to exceptional return, or necessarily after the `assume error`.

Reply:
From `assume error` in `f` , the analysis concludes that `f` cannot raise (because `List.hd` is not reachable), so `printf` line in `foo` is not reachable, as far as the analysis knowns.